[RFC] lantiq: add support for D-Link DWR-966

abajk opened the pull request:
Hardware:
- SoC: Lantiq GRX 389-720
- CPU: 2x MIPS 34Kc 600 MHz
- RAM: 128 MiB 300 MHz
- Flash: 128 MiB NAND
- Ethernet: Built-in Gigabit Ethernet switch, 4 ports
- WiFi 2.4G: Lantiq PSB 81312 b/g/n 2x2:2
- WiFi 5G: Qualcomm Atheros QCA9880 a/n/ac 3x3:3
- LTE: Cat6 Modem connected via ASM1042 PCI-E to USB 3.0 Hub
- USB: 2x USB 2.0
- LEDs: 5x RSSI, 4G, 3G, 2G, ANT, 5G, 2.4G, WPS, Internet, 2x PWR (Green and Red)
- Buttons: 1x Reset, 1x ANT, 1x WPS

Backup:
- enable ssh on stock firmware
- login via ssh:
 user: admin
 pass: admin
- mount USB drive
- make backup:
  $ nanddump --file /tmp/mounts/<directory>/dwr966.mtd1.nanddump /dev/mtd1
  $ nanddump --file /tmp/mounts/<directory>/dwr966.mtd2.nanddump /dev/mtd2
  $ nanddump --file /tmp/mounts/<directory>/dwr966.mtd4.nanddump /dev/mtd4

Installation:
- enable tftpserver on address 192.168.1.100 and connect ethernet cable to router
- copy initramfs to main tftp server directory and rename file to uImage:
  $ cp openwrt-lantiq-xrx300-dlink_dwr-966-initramfs-kernel.bin /srv/tftp/uImage
- connect via serial
- power on
- stop autoboot and enter:
  $ setenv loadaddr 0x84000000
  $ run net_ram
- download squashfs image via scp:
  $ scp openwrt-lantiq-xrx300-dlink_dwr-966-squashfs-sysupgrade.bin root@192.168.1.1:/tmp
- run sysupgrade
  $ sysupgrade /tmp/openwrt-lantiq-xrx300-dlink_dwr-966-squashfs-sysupgrade.bin
[At this momet boot fail 6 times due Invalid CRC]
- in GRX390 console enter (red_led is executed by u-boot if CRC is invalid):
  $ setenv red_led run flash_flash
  $ saveenv
- power off and power on. It normaly boot from flash.

Restore original firmware
- connect and mount USB drive
- install nand-utils
- erase flash:
  $ flash_erase /dev/mtd1 0 0
  $ flash_erase /dev/mtd2 0 0
  $ flash_erase /dev/mtd4 0 0
- write stock firmware:
  $ nandwrite -p /dev/mtd1 dwr966.mtd1.nanddump
  $ nandwrite -p /dev/mtd2 dwr966.mtd2.nanddump
  $ nandwrite -p /dev/mtd4 dwr966.mtd4.nanddump

Known issues:
- U-boot fails to boot linux due ivalid CRC checksum. Kernel is larger than 2MB (hardcoded value in u-boot) and this causes invalid CRC calculation.
- PCI-E doesn't work (WiFi, LTE)

I see that other people are working on support for ar10 and xrx330. This is a good time to add basic support and not to duplicate some of the work in the future.

Bootlog:
```
ROM VER: 1.0.0
CFG 06
NAND

ROM VER: 1.0.0
CFG 06
NAND

DDR autotuning Rev 1.0
DDR size from 0xa0000000 - 0xa7ffffff
DDR check ok... start booting...



U-Boot 2010.06-dirty-LANTIQ-v-2.3.08 (Aug 14 2015 - 10:30:43)

CLOCK CPU 600M RAM 300M
16 Bit RAM
DRAM:  128 MiB
NAND:  ONFI flash detected
ONFI param page 0 valid
NAND device: Manufacturer ID: 0x01, Chip ID: 0xf1 (AMD S34ML01G1)
128 MiB
Bad block table found at page 65472, version 0x01
Bad block table found at page 65408, version 0x01
In:    serial
Out:   serial
Err:   serial
Net:   internal phy using 25Mhz clock
Internal phy firmware version: 0x8434
ar10 Switch

Type "run flash_nfs" to mount root filesystem over NFS

part1_err=3,part2_err=3
Both partition firmware are CRC error, stop booting..
Hit any key to stop autoboot:  0

Creating 1 MTD partitions on "nand0":
0x0000001c0000-0x000007dc0000 : "mtd=4"
UBI: attaching mtd1 to ubi0
UBI: physical eraseblock size:   131072 bytes (128 KiB)
UBI: logical eraseblock size:    129024 bytes
UBI: smallest flash I/O unit:    2048
UBI: sub-page size:              512
UBI: VID header offset:          512 (aligned 512)
UBI: data offset:                2048
UBI: attached mtd1 to ubi0
UBI: MTD device name:            "mtd=4"
UBI: MTD device size:            124 MiB
UBI: number of good PEBs:        992
UBI: number of bad PEBs:         0
UBI: max. allowed volumes:       128
UBI: wear-leveling threshold:    4096
UBI: number of internal volumes: 1
UBI: number of user volumes:     8
UBI: available PEBs:             11
UBI: total number of reserved PEBs: 981
UBI: number of PEBs reserved for bad PEB handling: 9
UBI: max/mean erase counter: 22/10
Volume kernel found at volume id 0
read 0 bytes from volume 0 to 80800000(buf address)
Read [2580480] bytes
## Booting kernel from Legacy Image at 80800000 ...
   Image Name:   MIPS OpenWrt Linux-5.4.46
   Created:      2020-06-20  14:52:06 UTC
   Image Type:   MIPS Linux Kernel Image (lzma compressed)
   Data Size:    2487012 Bytes = 2.4 MiB
   Load Address: 80002000
   Entry Point:  80002000
   Verifying Checksum ... OK
   Uncompressing Kernel Image ... OK

Starting kernel ...

[    0.000000] Linux version 5.4.46 (aleksander@LAPTOP-OLEK) (gcc version 8.4.0 (OpenWrt GCC 8.4.0 r13605-2d1afa9278)) #0 SMP Sat Jun 20 14:52:06 2020
[    0.000000] SoC: xRX330 rev 1.1
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019556 (MIPS 34Kc)
[    0.000000] MIPS: machine is D-Link DWR-966
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Detected 1 available secondary CPU(s)
[    0.000000] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, VIPT, cache aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] percpu: Embedded 14 pages/cpu s26160 r8192 d22992 u57344
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 32480
[    0.000000] Kernel command line: console=ttyLTQ0,115200
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes, linear)
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes, linear)
[    0.000000] Writing ErrCtl register=0007fff0
[    0.000000] Readback ErrCtl register=0007fff0
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 120476K/131072K available (5798K kernel code, 218K rwdata, 1668K rodata, 1260K init, 220K bss, 10596K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=2, Nodes=1
[    0.000000] rcu: Hierarchical RCU implementation.
[    0.000000] rcu: RCU calculated value of scheduler-enlistment delay is 25 jiffies.
[    0.000000] NR_IRQS: 256
[    0.000000] random: get_random_bytes called from start_kernel+0x33c/0x54c with crng_init=0
[    0.000000] CPU Clock: 600MHz
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 6370868154 ns
[    0.000009] sched_clock: 32 bits at 300MHz, resolution 3ns, wraps every 7158278654ns
[    0.007881] Calibrating delay loop... 397.82 BogoMIPS (lpj=795648)
[    0.045921] pid_max: default: 32768 minimum: 301
[    0.050800] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.058016] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.068831] rcu: Hierarchical SRCU implementation.
[    0.074233] smp: Bringing up secondary CPUs ...
[    0.079698] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    0.079712] Primary data cache 32kB, 4-way, VIPT, cache aliases, linesize 32 bytes
[    0.079823] CPU1 revision is: 00019556 (MIPS 34Kc)
[    0.106834] Synchronize counters for CPU 1: done.
[    0.131152] smp: Brought up 1 node, 2 CPUs
[    0.139363] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645041785100000 ns
[    0.148993] futex hash table entries: 512 (order: 2, 16384 bytes, linear)
[    0.156145] pinctrl core: initialized pinctrl subsystem
[    0.162983] NET: Registered protocol family 16
[    0.185448] pinctrl-xway 1e100b10.pinmux: Init done
[    0.192032] dma-xway 1e104100.dma: Init done - hw rev: 8, ports: 5, channels: 24
[    0.240580] gpio-stp-xway 1e100bb0.stp: Init done
[    0.245924] usbcore: registered new interface driver usbfs
[    0.251405] usbcore: registered new interface driver hub
[    0.256894] usbcore: registered new device driver usb
[    0.265127] clocksource: Switched to clocksource MIPS
[    0.272378] NET: Registered protocol family 2
[    0.277673] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 6144 bytes, linear)
[    0.285962] TCP established hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.293656] TCP bind hash table entries: 1024 (order: 1, 8192 bytes, linear)
[    0.300777] TCP: Hash tables configured (established 1024 bind 1024)
[    0.307341] UDP hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.313811] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.321262] NET: Registered protocol family 1
[    0.325542] PCI: CLS 0 bytes, default 32
[    0.330228] gptu: totally 6 16-bit timers/counters
[    0.335086] gptu: misc_register on minor 63
[    0.339228] gptu: succeeded to request irq 126
[    0.343708] gptu: succeeded to request irq 127
[    0.348230] gptu: succeeded to request irq 128
[    0.352740] gptu: succeeded to request irq 129
[    0.357259] gptu: succeeded to request irq 130
[    0.361770] gptu: succeeded to request irq 131
[    0.366416] No VPEs reserved for AP/SP, not initialize VPE loader
[    0.366416] Pass maxvpes=<n> argument as kernel argument
[    0.373140] random: fast init done
[    0.377797] No TCs reserved for AP/SP, not initializing RTLX.
[    0.377797] Pass maxtcs=<n> argument as kernel argument
[    0.394834] workingset: timestamp_bits=14 max_order=15 bucket_order=1
[    0.412412] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.418176] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.440504] 1e100c00.serial: ttyLTQ0 at MMIO 0x1e100c00 (irq = 112, base_baud = 0) is a lantiq,asc
[    0.449404] printk: console [ttyLTQ0] enabled
[    0.449404] printk: console [ttyLTQ0] enabled
[    0.458070] printk: bootconsole [early0] disabled
[    0.458070] printk: bootconsole [early0] disabled
[    0.470603] nand: device found, Manufacturer ID: 0x01, Chip ID: 0xf1
[    0.475534] nand: AMD/Spansion S34ML01G1
[    0.479396] nand: 128 MiB, SLC, erase size: 128 KiB, page size: 2048, OOB size: 64
[    0.486926] Scanning device for bad blocks
[    0.591986] 8 fixed-partitions partitions found on MTD device 14000000.nand
[    0.597494] Creating 8 MTD partitions on "14000000.nand":
[    0.602874] 0x000000000000-0x000000100000 : "uboot"
[    0.609518] 0x000000100000-0x000000140000 : "ubootconfigA"
[    0.615412] 0x000000140000-0x000000180000 : "ubootconfigB"
[    0.621329] 0x000000180000-0x0000001c0000 : "gphyfirmware"
[    0.627133] 0x0000001c0000-0x000007dc0000 : "ubi"
[    0.633301] 0x000007dc0000-0x000007ec0000 : "calibration"
[    0.639073] 0x000007ec0000-0x000007fc0000 : "factory"
[    0.644497] 0x000007fc0000-0x000008000000 : "res"
[    0.652918] libphy: Fixed MDIO Bus: probed
[    0.666569] NET: Registered protocol family 10
[    0.674821] Segment Routing with IPv6
[    0.677204] NET: Registered protocol family 17
[    0.681553] 8021q: 802.1Q VLAN Support v1.8
[    0.798212] libphy: lantiq,xrx200-mdio: probed
[    0.807512] Intel XWAY PHY11G (xRX v1.2 integrated) 0:01: attached PHY driver [Intel XWAY PHY11G (xRX v1.2 integrated)] (mii_bus:phy_addr=0:01, irq=POLL)
[    0.821184] Intel XWAY PHY11G (xRX v1.2 integrated) 0:02: attached PHY driver [Intel XWAY PHY11G (xRX v1.2 integrated)] (mii_bus:phy_addr=0:02, irq=POLL)
[    0.834692] Intel XWAY PHY11G (xRX v1.2 integrated) 0:03: attached PHY driver [Intel XWAY PHY11G (xRX v1.2 integrated)] (mii_bus:phy_addr=0:03, irq=POLL)
[    0.848368] Intel XWAY PHY11G (xRX v1.2 integrated) 0:04: attached PHY driver [Intel XWAY PHY11G (xRX v1.2 integrated)] (mii_bus:phy_addr=0:04, irq=POLL)
[    0.863125] UBI: auto-attach mtd4
[    0.864986] ubi0: attaching mtd4
[    1.001894] ubi0: scanning is finished
[    1.035979] ubi0: attached mtd4 (name "ubi", size 124 MiB)
[    1.040077] ubi0: PEB size: 131072 bytes (128 KiB), LEB size: 129024 bytes
[    1.046859] ubi0: min./max. I/O unit sizes: 2048/2048, sub-page size 512
[    1.053509] ubi0: VID header offset: 512 (aligned 512), data offset: 2048
[    1.060255] ubi0: good PEBs: 992, bad PEBs: 0, corrupted PEBs: 0
[    1.066203] ubi0: user volume: 8, internal volumes: 1, max. volumes count: 128
[    1.073378] ubi0: max/mean erase counter: 22/10, WL threshold: 4096, image sequence number: 0
[    1.081844] ubi0: available PEBs: 0, total reserved PEBs: 992, PEBs reserved for bad PEB handling: 20
[    1.091073] ubi0: background thread "ubi_bgt0d" started, PID 510
[    1.096870] block ubiblock0_1: created from ubi0:1(rootfs)
[    1.102473] ubiblock: device ubiblock0_1 (rootfs) set to be root fil[    1.115918] VFS: Mounted root (squashfs filesystem) readonly on device 254:0.
[    1.125390] Freeing unused kernel memory: 1260K
[    1.128436] This architecture does not have kernel memory protection.
[    1.134867] Run /sbin/init as init process
[    1.778135] init: Console is alive
[    1.780400] init: - watchdog -
[    2.182878] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    2.284204] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    2.296384] SCSI subsystem initialized
[    2.303355] ehci-fsl: Freescale EHCI Host controller driver
[    2.308981] ehci-platform: EHCI generic platform driver
[    2.320509] dwc2 1e101000.usb: 1e101000.usb supply vusb_d not found, using dummy regulator
[    2.327531] dwc2 1e101000.usb: 1e101000.usb supply vusb_a not found, using dummy regulator
[    2.437537] dwc2 1e101000.usb: DWC OTG Controller
[    2.440816] dwc2 1e101000.usb: new USB bus registered, assigned bus number 1
[    2.447870] dwc2 1e101000.usb: irq 62, io mem 0x1e101000
[    2.454240] hub 1-0:1.0: USB hub found
[    2.456875] hub 1-0:1.0: 1 port detected
[    2.461992] dwc2 1e106000.usb: 1e106000.usb supply vusb_d not found, using dummy regulator
[    2.469026] dwc2 1e106000.usb: 1e106000.usb supply vusb_a not found, using dummy regulator
[    2.579391] dwc2 1e106000.usb: DWC OTG Controller
[    2.582724] dwc2 1e106000.usb: new USB bus registered, assigned bus number 2
[    2.589694] dwc2 1e106000.usb: irq 91, io mem 0x1e106000
[    2.596101] hub 2-0:1.0: USB hub found
[    2.598737] hub 2-0:1.0: 1 port detected
[    2.607783] usbcore: registered new interface driver usb-storage
[    2.612689] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    2.624709] init: - preinit -
[    3.387051] random: procd: uninitialized urandom read (4 bytes read)
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
[    4.377330] lantiq,xrx200-net 1e108000.eth eth0: port 3 got link
[    4.382065] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[    7.605179] UBIFS (ubi0:7): Mounting in unauthenticated mode
[    7.609762] UBIFS (ubi0:7): background thread "ubifs_bgt0_7" started, PID 608
[    7.725636] UBIFS (ubi0:7): UBIFS: mounted UBI device 0, volume 7, name "rootfs_data"
[    7.732040] UBIFS (ubi0:7): LEB size: 129024 bytes (126 KiB), min./max. I/O unit sizes: 2048 bytes/2048 bytes
[    7.741864] UBIFS (ubi0:7): FS size: 76511232 bytes (72 MiB, 593 LEBs), journal size 3870720 bytes (3 MiB, 30 LEBs)
[    7.752236] UBIFS (ubi0:7): reserved for root: 3613814 bytes (3529 KiB)
[    7.758820] UBIFS (ubi0:7): media format: w4/r0 (latest is w5/r0), UUID 31C136CD-EC32-40A1-90C1-1F9C393DC1F2, small LPT model
[    7.771219] mount_root: overlay filesystem has not been fully initialized yet
[    7.778154] mount_root: switching to ubifs overlay
[    7.805840] overlayfs: upper fs does not support xattr, falling back to index=off and metacopy=off.
- config restore -
[    7.907447] urandom-seed: Seed file not found (/etc/urandom.seed)
[    8.083845] lantiq,xrx200-net 1e108000.eth eth0: port 3 lost link
[    8.101597] procd: - early -
[    8.103164] procd: - watchdog -
[    8.742648] procd: - watchdog -
[    8.744856] procd: - ubus -
[    8.815333] random: ubusd: uninitialized urandom read (4 bytes read)
[    8.821783] random: ubusd: uninitialized urandom read (4 bytes read)
[    8.827248] random: ubusd: uninitialized urandom read (4 bytes read)
[    8.834915] procd: - init -
Please press Enter to activate this console.
[    9.620449] kmodloader: loading kernel modules from /etc/modules.d/*
[    9.678586] Infineon Technologies DEU driver version 2.0.0
[    9.847519] IFX DEU DES initialized (multiblock).
[    9.978345] urngd: v1.0.2 started.
[    9.990004] IFX DEU AES initialized (multiblock).
[   10.048074] IFX DEU ARC4 initialized (multiblock).
[   10.079065] IFX DEU SHA1 initialized.
[   10.108635] IFX DEU MD5 initialized.
[   10.138657] IFX DEU SHA1_HMAC initialized.
[   10.168748] IFX DEU MD5_HMAC initialized.
[   10.199947] xt_time: kernel timezone is -0000
[   10.233624] PPP generic driver version 2.4.2
[   10.240501] NET: Registered protocol family 24
[   10.264501] kmodloader: done loading kernel modules from /etc/modules.d/*
[   10.298663] random: crng init done
[   10.300595] random: 6 urandom warning(s) missed due to ratelimiting
[   20.188929] device eth0 entered promiscuous mode
[   20.196879] br-lan: port 1(eth0.1) entered blocking state
[   20.200864] br-lan: port 1(eth0.1) entered disabled state
[   20.206759] device eth0.1 entered promiscuous mode
[   23.221438] lantiq,xrx200-net 1e108000.eth eth0: port 3 got link
[   23.226230] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[   23.239016] br-lan: port 1(eth0.1) entered blocking state
[   23.243008] br-lan: port 1(eth0.1) entered forwarding state
[   23.252339] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready




BusyBox v1.31.1 () built-in shell (ash)

  _______                     ________        __
 |       |.-----.-----.-----.|  |  |  |.----.|  |_
 |   -   ||  _  |  -__|     ||  |  |  ||   _||   _|
 |_______||   __|_____|__|__||________||__|  |____|
          |__| W I R E L E S S   F R E E D O M
 -----------------------------------------------------
 OpenWrt SNAPSHOT, r13605-2d1afa9278
 -----------------------------------------------------
=== WARNING! =====================================
There is no root password defined on this device!
Use the "passwd" command to set up a new password
in order to prevent unauthorized SSH logins.
--------------------------------------------------
root@OpenWrt:/# 
root@OpenWrt:/# cat /proc/cpuinfo 
system type             : xRX330 rev 1.1
machine                 : D-Link DWR-966
processor               : 0
cpu model               : MIPS 34Kc V5.6
BogoMIPS                : 397.82
wait instruction        : yes
microsecond timers      : yes
tlb_entries             : 16
extra interrupt vector  : yes
hardware watchpoint     : yes, count: 4, address/irw mask: [0x0ffc, 0x0ffc, 0x0ffb, 0x0ffb]
isa                     : mips1 mips2 mips32r1 mips32r2 mips64r1 mips64r2
ASEs implemented        : mips16 dsp mt
Options implemented     : tlb 4kex 4k_cache 32fpr prefetch mcheck ejtag llsc dc_aliases userlocal vint perf_cntr_intr_bit perf
shadow register sets    : 1
kscratch registers      : 0
package                 : 0
core                    : 0
VPE                     : 0
VCED exceptions         : not available
VCEI exceptions         : not available

processor               : 1
cpu model               : MIPS 34Kc V5.6
BogoMIPS                : 400.38
wait instruction        : yes
microsecond timers      : yes
tlb_entries             : 16
extra interrupt vector  : yes
hardware watchpoint     : yes, count: 4, address/irw mask: [0x0ffc, 0x0ffc, 0x0ffb, 0x0ffb]
isa                     : mips1 mips2 mips32r1 mips32r2 mips64r1 mips64r2
ASEs implemented        : mips16 dsp mt
Options implemented     : tlb 4kex 4k_cache 32fpr prefetch mcheck ejtag llsc dc_aliases userlocal vint perf_cntr_intr_bit perf
shadow register sets    : 1
kscratch registers      : 0
package                 : 0
core                    : 0
VPE                     : 1
VCED exceptions         : not available
VCEI exceptions         : not available

root@OpenWrt:/#
root@OpenWrt:/# cat /proc/mtd 
dev:    size   erasesize  name
mtd0: 00100000 00020000 "uboot"
mtd1: 00040000 00020000 "ubootconfigA"
mtd2: 00040000 00020000 "ubootconfigB"
mtd3: 00040000 00020000 "gphyfirmware"
mtd4: 07c00000 00020000 "ubi"
mtd5: 00100000 00020000 "calibration"
mtd6: 00100000 00020000 "factory"
mtd7: 00040000 00020000 "res"
root@OpenWrt:/# 
root@OpenWrt:/# ping 192.168.1.100
PING 192.168.1.100 (192.168.1.100): 56 data bytes
64 bytes from 192.168.1.100: seq=0 ttl=64 time=0.867 ms
64 bytes from 192.168.1.100: seq=1 ttl=64 time=0.607 ms
64 bytes from 192.168.1.100: seq=2 ttl=64 time=0.573 ms
^C
--- 192.168.1.100 ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 0.573/0.682/0.867 ms
root@OpenWrt:/#
root@OpenWrt:/# lsusb
Bus 002 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
root@OpenWrt:/# [  250.038926] usb 2-1: new high-speed USB device number 2 using dwc2
[  250.260882] usb-storage 2-1:1.0: USB Mass Storage device detected
[  250.268297] scsi host0: usb-storage 2-1:1.0
[  251.291963] scsi 0:0:0:0: Direct-Access     Kingston DataTraveler 2.0 1.00 PQ: 0 ANSI: 2
[  251.302768] sd 0:0:0:0: [sda] 15679488 512-byte logical blocks: (8.03 GB/7.48 GiB)
[  251.309512] sd 0:0:0:0: [sda] Write Protect is off
[  251.314597] sd 0:0:0:0: [sda] No Caching mode page found
[  251.319011] sd 0:0:0:0: [sda] Assuming drive cache: write through
[  251.448711]  sda: sda1
[  251.454208] sd 0:0:0:0: [sda] Attached SCSI removable disk

root@OpenWrt:/# 
root@OpenWrt:/# lsusb
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 002 Device 002: ID 0951:1603 Kingston Technology DataTraveler 1GB/2GB Pen Drive
Bus 002 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
root@OpenWrt:/# 
```

TODO:
- [x] Rebase to current master
- [x] Add changes suggested by adschm
- [x] Replace gphy clock alias patch with upstream version
- [x] Add upstream patch for STP
- [x] Dropped unused and untested nodes from xrx330.dtsi (vmmc, spi, usif, mei and ppe)
- [x] Dropped unused labels from dts
- [x] Switch to 5.10
- [ ] Fix RX hang under iperf and switch to DSA driver
- [ ] Extract GPHY firmware from the flash partition

abajk committed:
lantiq: add support for D-Link DWR-966

Hardware:
- SoC: Lantiq GRX 389-720
- CPU: 2x MIPS 34Kc 600 MHz
- RAM: 128 MiB 300 MHz
- Flash: 128 MiB NAND
- Ethernet: Built-in Gigabit Ethernet switch, 4 ports
- WiFi 2.4G: Lantiq PSB 81312 b/g/n 2x2:2
- WiFi 5G: Qualcomm Atheros QCA9880 a/n/ac 3x3:3
- LTE: Cat6 Modem connected via ASM1042 PCI-E to USB 3.0 Hub
- USB: 2x USB 2.0
- LEDs: 5x RSSI, 4G, 3G, 2G, ANT, 5G, 2.4G, WPS, Internet, 2x PWR (Green and Red)
- Buttons: 1x Reset, 1x ANT, 1x WPS

Backup:
- enable ssh on stock firmware
- login via ssh:
 user: admin
 pass: admin
- mount USB drive
- make backup:
  $ nanddump --file /tmp/mounts/<directory>/dwr966.mtd1.nanddump /dev/mtd1
  $ nanddump --file /tmp/mounts/<directory>/dwr966.mtd2.nanddump /dev/mtd2
  $ nanddump --file /tmp/mounts/<directory>/dwr966.mtd4.nanddump /dev/mtd4

Installation:
- enable tftpserver on address 192.168.1.100 and connect ethernet cable to router
- copy initramfs to main tftp server directory and rename file to uImage:
  $ cp openwrt-lantiq-xrx200-dlink_dwr-966-initramfs-kernel.bin /srv/tftp/uImage
- connect via serial
- power on
- stop autoboot and enter:
  $ setenv loadaddr 0x84000000
  $ run net_ram
- download squashfs image via scp:
  $ scp openwrt-lantiq-xrx200-dlink_dwr-966-squashfs-sysupgrade.bin root@192.168.1.1:/tmp
- run sysupgrade
  $ sysupgrade /tmp/openwrt-lantiq-xrx200-dlink_dwr-966-squashfs-sysupgrade.bin
[At this momet boot fail 6 times due Invalid CRC]
- in GRX390 console enter (red_led is executed by u-boot if CRC is invalid):
  $ setenv red_led run flash_flash
  $ saveenv
- power off and power on. It normaly boot from flash.

Restore original firmware
- connect and mount USB drive
- install nand-utils
- erase flash:
  $ flash_erase /dev/mtd1 0 0
  $ flash_erase /dev/mtd2 0 0
  $ flash_erase /dev/mtd4 0 0
- write stock firmware:
  $ nandwrite -p /dev/mtd1 dwr966.mtd1.nanddump
  $ nandwrite -p /dev/mtd2 dwr966.mtd2.nanddump
  $ nandwrite -p /dev/mtd4 dwr966.mtd4.nanddump

Known issues:
- U-boot fails to boot linux due ivalid CRC checksum. Kernel is larger than 2MB (hardcoded value in u-boot) and this causes invalid CRC calculation.
- PCI-E doesn't work (WiFi, LTE)

Signed-off-by: Aleksander Jan Bajkowski <olek2@wp.pl>

abajk committed:
lantiq-gphy-firmware: add firmware for xrx300 and xrx330

They are from DWR-966 GPL sources[1].

[1] https://github.com/brunompena/dwr-966/tree/master/target/linux/lantiq/files/firmware/lantiq
Signed-off-by: Aleksander Jan Bajkowski <olek2@wp.pl>

adrianschmutzler commented:
> lantiq: add initial support for D-Link DWR-966

Why "initial"?

adrianschmutzler commented:
From the commit message:

> Restore original firmware
> -

There seems to be an empty bullet point here

> Known issues:
> - U-boot fails on boot due incorrect CRC checksum, but manualy boot from flash

Had some problems here, do you mean "but manually boots from flash"?

adrianschmutzler commented on target/linux/lantiq/files-4.19/arch/mips/boot/dts/lantiq/grx390.dtsi:
```diff
@@ -0,0 +1,321 @@
+#include <dt-bindings/gpio/gpio.h>
```
please add an SPDX license identifier.

adrianschmutzler commented on target/linux/lantiq/files-4.19/arch/mips/boot/dts/lantiq/grx390_dlink_dwr-966.dts:
```diff
@@ -0,0 +1,288 @@
+/dts-v1/;
```
please add an SPDX license identifier.

adrianschmutzler commented on target/linux/lantiq/files-4.19/arch/mips/boot/dts/lantiq/grx390_dlink_dwr-966.dts:
```diff
@@ -0,0 +1,288 @@
+			gpios = <&stp 23 GPIO_ACTIVE_LOW>;
+		};
+		web_green: web-green {
+			label = "green:internet";
```
common scheme for label is "model:color:use", e.g. "dwr-966:green:internet" here ...

adrianschmutzler commented on target/linux/lantiq/files-4.19/arch/mips/boot/dts/lantiq/grx390_dlink_dwr-966.dts:
```diff
@@ -0,0 +1,288 @@
+			label = "green:wps";
+			gpios = <&stp 23 GPIO_ACTIVE_LOW>;
+		};
+		web_green: web-green {
```
The node name should match the last part of the label, unless there are multiple leds which only deviate in color.

So, this would be just "internet", if there were multiple internet LEDs those would be "internet_green", "internet_red" etc.

The DT label (prefix in this line) will then have the same string prefixed by "led_", as this DT label is used on global context. So, all put together:

```
   led_internet: internet {
      label = "dwr-966:green:internet";
      gpios = <&stp 22 GPIO_ACTIVE_LOW>;
   };
```

However, since the DT label does not seem to be used anywhere, you could drop it and simply remove the "led_internet:" for this node.

Despite, please add an empty line between the led nodes.

adrianschmutzler commented on target/linux/lantiq/files-4.19/arch/mips/boot/dts/lantiq/grx390_dlink_dwr-966.dts:
```diff
@@ -0,0 +1,288 @@
+			};
+			partition@01c0000 {
+				label = "ubi";
+				reg = <0x01c0000 0x7c00000>;
```
Please be consistent on the leading zeros for the partition location and size. Either use no leading zeros at all, or have all addresses for all partition have the same number of total digits.

adrianschmutzler commented on target/linux/lantiq/files-4.19/arch/mips/boot/dts/lantiq/grx390_dlink_dwr-966.dts:
```diff
@@ -0,0 +1,288 @@
+};
+
+&eth0 {
+	lan: interface@0 {
```
Is this label required somewhere?

olek210 replied:
dropped

adrianschmutzler commented on target/linux/lantiq/files-4.19/arch/mips/boot/dts/lantiq/grx390_dlink_dwr-966.dts:
```diff
@@ -0,0 +1,288 @@
+
+		nand_out {
+			lantiq,groups = "nand cle", "nand ale", "nand rd",
+						"nand d1",
```
That indent does not really look helpful.

olek210 replied:
fixed

adrianschmutzler commented on target/linux/lantiq/files-4.19/arch/mips/boot/dts/lantiq/grx390_dlink_dwr-966.dts:
```diff
@@ -0,0 +1,288 @@
+};
+
+&stp {
+	status = "ok";
```
please use "okay" instead of "ok" on all statusses.

adrianschmutzler commented on target/linux/lantiq/files-5.4/arch/mips/boot/dts/lantiq/grx390.dtsi:
```diff
@@ -0,0 +1,321 @@
+#include <dt-bindings/gpio/gpio.h>
```
All comments from 4.19 apply to 5.4 files as well.

adrianschmutzler commented on target/linux/lantiq/image/Makefile:
```diff
@@ -130,6 +130,11 @@ include tp-link.mk
+ifeq ($(SUBTARGET),xrx300)
+DEFAULT_SOC := grx390
+include grx390.mk
+endif
```
Do we really need another subtarget here?

olek210 replied:
> Do we really need another subtarget here?

Lantiq SDK split xrx200 and xrx300 into separate subtargets.
Differences:
- newer builtin WiFi (WAVE400 vs WAVE300)
- gphy in ar10 switch use another firmware
- newer vdsl (VRX318)
- 3x PCIe vs 1xPCI+1xPCIe (OpenWRT PCIe driver didn't work with grx390 and I have plan to use mainline dw-pcie driver)

At this point, I can merge in with xrx200. There may be a problem with this in the future.

hauke replied:
* The xrx200 does not have any wifi integrated, I think a wave300 is integrated in the xrx300/ xrx330 family, wave400 does not exist (unlucky number)
* The VRX318 is an external DSL PCIe chip
* The NAND core is the same as in the xrx200 just some more extensions. 

The GRX350 / 550 is a different chip with a InterAptiv CPU.

Here is a recent NAND driver:
https://lkml.org/lkml/2020/6/16/302
It is the same NAND controller as in the xrx300, you just have an older version, it could be that it does not support all features.

A PCIe controller driver is also upstream:
https://git.kernel.org/linus/ed22aaaede44f647477a5048e62855c0ed49c9bd
I am not sure if the xrx300 uses a Lantiq PCIe PHY like the xrx200 or a Synopsys PCIe PHY like the xrx500 and LGM. If it uses a synopsys PCIe PHY it should be easy to make this driver work.
The xrx200 PCI (without express) controller driver is stil using the old MIPS PCIe driver framework which is a problem.

olek210 replied:
Thanks
I didn't know that they are almost identical. I moved it to the xrx200 subtarget.

adrianschmutzler commented on target/linux/lantiq/image/grx390.mk:
```diff
@@ -0,0 +1,9 @@
+  $(Device/NAND)
+  DEVICE_VENDOR := D-Link
+  DEVICE_MODEL := DWR-966
+  BOARD_NAME := DWR-966
```
Is BOARD_NAME really required or can it be dropped?

olek210 replied:
dropped

adrianschmutzler commented on target/linux/lantiq/image/grx390.mk:
```diff
@@ -0,0 +1,9 @@
+  DEVICE_MODEL := DWR-966
+  BOARD_NAME := DWR-966
+  DEVICE_PACKAGES := kmod-usb-dwc2
+  SUPPORTED_DEVICES += DWR-966
```
This line can be dropped.

adrianschmutzler commented on target/linux/lantiq/xrx300/base-files/etc/board.d/02_network:
```diff
@@ -0,0 +1,47 @@
+
+. /lib/functions/uci-defaults.sh
+. /lib/functions/system.sh
+. /lib/functions/lantiq.sh
```
This line can be dropped, as you don't use the functions in the file.

adrianschmutzler commented on target/linux/lantiq/xrx300/base-files/lib/upgrade/platform.sh:
```diff
@@ -0,0 +1,19 @@
+PART_NAME=firmware
```
This file doesn't need the execute bit.

adrianschmutzler commented:
> lantiq: refresh patches

I'd squash this into the patch adding the extra patches.

olek210 commented:
Thanks. I will fix it :)
> Had some problems here, do you mean "but manually boots from flash"?

There is still a problem with the u-boot. U-Boot stops booting due invalid CRC:
```
part1_err=3,part2_err=3
Both partition firmware are CRC error, stop booting..
Hit any key to stop autoboot:  0
```

After that it boots normally:
```
GRX330 # run flash_flash
```

Kernel works fine, and patch for uboot-lantiq will be in the next PR.

olek210 commented:
@adrianschmutzler 
Fixed and ready to merge :)

olek210 commented:
Rebased. Commit "lantiq: refresh patches" did the same as ef9cee49807a360c99abfc2bbf0ab873619d6f11.

hauke commented on target/linux/lantiq/files-4.19/arch/mips/boot/dts/lantiq/grx390.dtsi:
```diff
@@ -0,0 +1,322 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
```
Please rename this file to xrx300.dtsi or xrx330.dtsi 
The name grx390 was never used in production

olek210 replied:
renamed to xrx330

hauke commented on target/linux/lantiq/files-4.19/arch/mips/boot/dts/lantiq/grx390.dtsi:
```diff
@@ -0,0 +1,322 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+#include <dt-bindings/gpio/gpio.h>
+
```
No need to add this for kernel 4.19 any more, it should be sufficient if this is only working with kernel 5.4

abajk committed:
lantiq: xrx200 set soc version

xrx200 subtarget has now support for xrx300 and vr9 SoCs, so set SoC version for every board.

Signed-off-by: Aleksander Jan Bajkowski <olek2@wp.pl>

hauke commented on target/linux/lantiq/xrx200/config-5.4:
```diff
@@ -16,7 +16,7 @@ CONFIG_CRYPTO_MANAGER=y
 CONFIG_CRYPTO_NULL2=y
 CONFIG_CRYPTO_ZSTD=y
-CONFIG_EXTRA_FIRMWARE="lantiq/xrx200_phy11g_a14.bin lantiq/xrx200_phy11g_a22.bin lantiq/xrx200_phy22f_a14.bin lantiq/xrx200_phy22f_a22.bin"
+CONFIG_EXTRA_FIRMWARE="lantiq/xrx200_phy11g_a14.bin lantiq/xrx200_phy11g_a22.bin lantiq/xrx200_phy22f_a14.bin lantiq/xrx200_phy22f_a22.bin lantiq/xrx300_phy11g_a21.bin lantiq/xrx300_phy22f_a21.bin"
```
Adding the GPHY FW here is problematic.
Under which license did you receive this firmware?

The GPHY Firmware is distributed here for example:
https://intel.prpl.dev/intel/ltq_fw_phy_ip 
This the license does not allow a redistribution.

We could for example extract it from the flash partition at runtime and load the firmware after the kernel booted up. This should be legally save.

olek210 replied:
Whole [DWR966 firmware](https://github.com/brunompena/dwr-966) is licensed under GPL.
I don't see anywhere that gphy firmware is under a different license.

It will be easier to convert GPHY firmware to a package like in the dsl firmware and load the switch driver after rootfs mounting. Is this acceptable?

hauke replied:
With the DSA switch driver it is possible to load the GPHY firmware later from user space.
Should we wait till that gets merged?

abajk replied:
I am planning to switch to a DSA driver. Previously, DSA driver crashed after a few seconds under iperf. This weekend I will test this driver with the latest patches.

adrianschmutzler commented on target/linux/lantiq/image/xrx330.mk:
```diff
@@ -0,0 +1,17 @@
+		-kmod-ltq-vdsl-vr9-mei \
+		-kmod-ltq-vdsl-vr9 \
+		-kmod-ltq-atm-vr9 \
+		-kmod-ltq-ptm-vr9 \
```
You are not removing kmod-ltq-deu-vr9 here, although it's also selected in xrx200/target.mk. Is this intended?

olek210 replied:
Yes, DEU is a hardware cryptographic accelerator and is built in xrx300.

adschm commented on target/linux/lantiq/files-5.4/arch/mips/boot/dts/lantiq/xrx330.dtsi:
```diff
@@ -0,0 +1,322 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
```
lantiq uses files instead of files-5.4 now. Apart from that, I've recently moved the dts-v1 statement to the dtsi files; so, I'd be happy if you could move it from DTS to DTSI as well.

adschm replied:
BTW GPL-2.0+ is deprecated, the proper term is GPL-2.0-or-later

adschm commented on target/linux/lantiq/files-5.4/arch/mips/boot/dts/lantiq/xrx330_dlink_dwr-966.dts:
```diff
@@ -0,0 +1,295 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
```
GPL-2.0-or-later

adschm commented on target/linux/lantiq/files-5.4/arch/mips/boot/dts/lantiq/xrx330_dlink_dwr-966.dts:
```diff
@@ -0,0 +1,295 @@
+			label = "reset";
+			gpios = <&gpio 9 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
```
I'd prefer empty lines between the nodes (generally).

adschm commented on target/linux/lantiq/files-5.4/arch/mips/boot/dts/lantiq/xrx330_dlink_dwr-966.dts:
```diff
@@ -0,0 +1,295 @@
+		compatible = "gpio-leds";
+
+		wps {
+			label = "dwr-966:green:wps";
```
The model prefix (dwr-966:) has been dropped from LED labels, so you can just use "green:wps" etc.

adschm commented on target/linux/lantiq/image/xrx330.mk:
```diff
@@ -0,0 +1,17 @@
+		-ltq-vdsl-app \
+		-dsl-vrx200-firmware-xdsl-a \
+		-dsl-vrx200-firmware-xdsl-b-patch \
+		-ppp-mod-pppoa
```
Hanging indent typically is one tab in these files.

hauke commented:
The DSA driver from  #3085  was just merge, could you please rebase this on top of current master.
